File: src/main/java/no/nav/tjenestepensjon/simulering/service/SimpleSimuleringService.java
package no.nav.tjenestepensjon.simulering.service;

import static java.util.stream.Collectors.toList;

import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_NAME;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_FEIL;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_MANGEL;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_OK;
import static no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_UFUL;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import no.nav.tjenestepensjon.simulering.AppMetrics;
import no.nav.tjenestepensjon.simulering.AsyncExecutor;
import no.nav.tjenestepensjon.simulering.AsyncExecutor.AsyncResponse;
import no.nav.tjenestepensjon.simulering.TjenestepensjonsimuleringEndpointRouter;
import no.nav.tjenestepensjon.simulering.consumer.FindTpLeverandorCallable;
import no.nav.tjenestepensjon.simulering.consumer.TpConfigConsumer;
import no.nav.tjenestepensjon.simulering.consumer.TpRegisterConsumer;
import no.nav.tjenestepensjon.simulering.domain.TPOrdning;
import no.nav.tjenestepensjon.simulering.domain.TpLeverandor;
import no.nav.tjenestepensjon.simulering.exceptions.DuplicateStillingsprosentEndDateException;
import no.nav.tjenestepensjon.simulering.exceptions.MissingStillingsprosentException;
import no.nav.tjenestepensjon.simulering.exceptions.NoTpOrdningerFoundException;
import no.nav.tjenestepensjon.simulering.exceptions.SoapFaultException;
import no.nav.tjenestepensjon.simulering.exceptions.StillingsprosentCallableException;
import no.nav.tjenestepensjon.simulering.rest.IncomingRequest;
import no.nav.tjenestepensjon.simulering.rest.OutgoingResponse;
import no.nav.tjenestepensjon.simulering.rest.OutgoingResponse.SimulertPensjon;
import no.nav.tjenestepensjon.simulering.rest.SimuleringEndpoint;

@Service
public class SimpleSimuleringService implements SimuleringEndpoint.SimuleringService {

    private static final Logger LOG = LoggerFactory.getLogger(SimpleSimuleringService.class);

    private final TjenestepensjonsimuleringEndpointRouter simuleringEndPointRouter;
    private final StillingsprosentService stillingsprosentService;
    private final TpConfigConsumer tpConfigConsumer;
    private final List<TpLeverandor> tpLeverandorList;
    private final TpRegisterConsumer tpRegisterConsumer;
    private final AsyncExecutor<TpLeverandor, FindTpLeverandorCallable> asyncExecutor;
    private final AppMetrics metrics;

    public SimpleSimuleringService(TjenestepensjonsimuleringEndpointRouter simuleringEndPointRouter, StillingsprosentService stillingsprosentService,
            TpConfigConsumer tpConfigConsumer,
            List<TpLeverandor> tpLeverandorList, TpRegisterConsumer tpRegisterConsumer,
            AsyncExecutor<TpLeverandor, FindTpLeverandorCallable> asyncExecutor, AppMetrics metrics) {
        this.simuleringEndPointRouter = simuleringEndPointRouter;
        this.stillingsprosentService = stillingsprosentService;
        this.tpConfigConsumer = tpConfigConsumer;
        this.tpLeverandorList = tpLeverandorList;
        this.tpRegisterConsumer = tpRegisterConsumer;
        this.asyncExecutor = asyncExecutor;
        this.metrics = metrics;
    }

    @Override
    public OutgoingResponse simuler(IncomingRequest request) {
        OutgoingResponse response = new OutgoingResponse();
        StillingsprosentResponse stillingsprosentResponse = null;
        try {
            List<TPOrdning> tpOrdningList = tpRegisterConsumer.getTpOrdningerForPerson(request.getFnr());
            Map<TPOrdning, TpLeverandor> tpOrdningAndLeverandorMap = getTpLeverandorer(tpOrdningList);
            stillingsprosentResponse = stillingsprosentService.getStillingsprosentListe(request.getFnr(), tpOrdningAndLeverandorMap);

            if (stillingsprosentResponse.getTpOrdningStillingsprosentMap().size() == 0) {
                response.setSimulertPensjonListe(addResponseInfoWhenError("", "Could not get stillingsprosent from any TP-Providers"));
                return response;
            }

            TPOrdning tpOrdning = stillingsprosentService.getLatestFromStillingsprosent(stillingsprosentResponse.getTpOrdningStillingsprosentMap());
            TpLeverandor tpLeverandor = tpOrdningAndLeverandorMap.get(tpOrdning);
            List<SimulertPensjon> simulertPensjonList =
                    simuleringEndPointRouter.simulerPensjon(request, tpOrdning, tpLeverandor, stillingsprosentResponse.getTpOrdningStillingsprosentMap());
            response.setSimulertPensjonListe(addResponseInfoWhenSimulert(simulertPensjonList, stillingsprosentResponse));
        } catch (DuplicateStillingsprosentEndDateException e) {
            response.setSimulertPensjonListe(addResponseInfoWhenError("PARF", e.getMessage()));
        } catch (MissingStillingsprosentException | SoapFaultException e) {
            response.setSimulertPensjonListe(addResponseInfoWhenError("IKKE", e.getMessage()));
        } catch (NoTpOrdningerFoundException e) {
            response.setSimulertPensjonListe(addResponseInfoWhenError("", e.getMessage()));
        }
        return response;
    }

    private List<SimulertPensjon> addResponseInfoWhenError(String feilKode, String msg) {
        metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_FEIL);
        List<SimulertPensjon> simulertPensjonList = List.of(new SimulertPensjon());
        simulertPensjonList.forEach(simulertPensjon -> simulertPensjon.setFeilkode(feilKode));
        simulertPensjonList.forEach(simulertPensjon -> simulertPensjon.setFeilbeskrivelse(msg));
        simulertPensjonList.forEach(simulertPensjon -> simulertPensjon.setStatus("FEIL"));
        return simulertPensjonList;
    }

    private List<SimulertPensjon> addResponseInfoWhenSimulert(List<SimulertPensjon> simulertPensjonList, StillingsprosentResponse stillingsprosentResponse) {
        List<String> utelatteTpNr = stillingsprosentResponse.getExceptions().stream()
                .filter(e -> e.getCause() instanceof StillingsprosentCallableException)
                .map(e -> ((StillingsprosentCallableException) e.getCause()).getTpOrdning().getTpId())
                .collect(toList());
        List<String> inkluderteTpNr = stillingsprosentResponse.getTpOrdningStillingsprosentMap().keySet().stream()
                .map(TPOrdning::getTpId)
                .collect(toList());
        simulertPensjonList.forEach(simulertPensjon -> {
            simulertPensjon.setUtelatteTpnr(utelatteTpNr);
            simulertPensjon.setInkluderteTpnr(inkluderteTpNr);
            if (utelatteTpNr.size() > 0) {
                simulertPensjon.setStatus("UFUL");
            }
        });
        incrementMetrics(simulertPensjonList, utelatteTpNr);
        return simulertPensjonList;
    }

    private void incrementMetrics(List<SimulertPensjon> simulertPensjonList, List<String> utelatteTpNr) {
        boolean ufullstendig = utelatteTpNr.size() > 0;
        boolean mangelfull = simulertPensjonList.stream()
                .anyMatch(simulertPensjon -> simulertPensjon.getUtbetalingsperioder().stream()
                        .anyMatch(utbetalingsperiode -> utbetalingsperiode.getMangelfullSimuleringkode() != null));

        if (ufullstendig) {
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_UFUL);
        }
        if (mangelfull) {
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_MANGEL);
        }
        if (!ufullstendig && !mangelfull) {
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_OK);
        }
    }

    private Map<TPOrdning, TpLeverandor> getTpLeverandorer(List<TPOrdning> tpOrdningList) {
        Map<TPOrdning, FindTpLeverandorCallable> callableMap = new HashMap<>();
        tpOrdningList.forEach(tpOrdning -> callableMap.put(tpOrdning, new FindTpLeverandorCallable(tpOrdning, tpConfigConsumer, tpLeverandorList)));
        AsyncResponse<TPOrdning, TpLeverandor> asyncResponse = asyncExecutor.executeAsync(callableMap);
        return asyncResponse.getResultMap();
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/service/SimpleSimuleringService.kt
package no.nav.tjenestepensjon.simulering.service

import no.nav.tjenestepensjon.simulering.AppMetrics
import no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_NAME
import no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_FEIL
import no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_MANGEL
import no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_OK
import no.nav.tjenestepensjon.simulering.AppMetrics.Metrics.APP_TOTAL_SIMULERING_UFUL
import no.nav.tjenestepensjon.simulering.AsyncExecutor
import no.nav.tjenestepensjon.simulering.TjenestepensjonsimuleringEndpointRouter
import no.nav.tjenestepensjon.simulering.consumer.FindTpLeverandorCallable
import no.nav.tjenestepensjon.simulering.consumer.TpConfigConsumer
import no.nav.tjenestepensjon.simulering.consumer.TpRegisterConsumer
import no.nav.tjenestepensjon.simulering.domain.TpLeverandor
import no.nav.tjenestepensjon.simulering.exceptions.NoTpOrdningerFoundException
import no.nav.tjenestepensjon.simulering.exceptions.SimuleringException
import no.nav.tjenestepensjon.simulering.exceptions.StillingsprosentCallableException
import no.nav.tjenestepensjon.simulering.model.v1.domain.TPOrdning
import no.nav.tjenestepensjon.simulering.model.v1.request.SimulerPensjonRequest
import no.nav.tjenestepensjon.simulering.model.v1.response.SimulerOffentligTjenestepensjonResponse
import no.nav.tjenestepensjon.simulering.model.v1.response.SimulertPensjon
import no.nav.tjenestepensjon.simulering.model.v1.response.SimulertPensjonFeil
import no.nav.tjenestepensjon.simulering.model.v1.response.SimulertPensjonOK
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.Collections.singletonList
import java.util.concurrent.ExecutionException

@Service
class SimpleSimuleringService(
        private val simuleringEndPointRouter: TjenestepensjonsimuleringEndpointRouter,
        private val stillingsprosentService: StillingsprosentService,
        private val tpConfigConsumer: TpConfigConsumer,
        private val tpLeverandorList: List<TpLeverandor>,
        private val tpRegisterConsumer: TpRegisterConsumer,
        private val asyncExecutor: AsyncExecutor<TpLeverandor, FindTpLeverandorCallable>,
        private val metrics: AppMetrics
) : SimuleringService {
    override fun simulerOffentligTjenestepensjon(request: SimulerPensjonRequest) =
            SimulerOffentligTjenestepensjonResponse(
                    simulertPensjonListe = try {
                        val tpOrdningAndLeverandorMap = tpRegisterConsumer.getTpOrdningerForPerson(request.fnr)
                                .let(::getTpLeverandorer)
                        val stillingsprosentResponse = stillingsprosentService.getStillingsprosentListe(request.fnr, tpOrdningAndLeverandorMap)
                        stillingsprosentResponse.tpOrdningStillingsprosentMap
                                .ifEmpty { throw NoTpOrdningerFoundException("Could not get stillingsprosent from any TP-Providers") }
                                .let(stillingsprosentService::getLatestFromStillingsprosent)
                                .let { tpOrdning ->
                                    simuleringEndPointRouter.simulerPensjon(
                                            request = request,
                                            tpOrdning = tpOrdning,
                                            tpLeverandor = tpOrdningAndLeverandorMap[tpOrdning]!!,
                                            tpOrdningStillingsprosentMap = stillingsprosentResponse.tpOrdningStillingsprosentMap
                                    )
                                }.let { simulertPensjonList ->
                                    addResponseInfoWhenSimulert(
                                            simulertPensjonList,
                                            stillingsprosentResponse
                                    )
                                }
                    } catch (e: SimuleringException) {
                        addResponseInfoWhenError(e)
                    }
            )

    private fun addResponseInfoWhenError(e: SimuleringException): List<SimulertPensjon> {
        metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_FEIL)
        return singletonList(SimulertPensjonFeil(
                feilkode = e.feilkode,
                feilbeskrivelse = e.message,
                status = "FEIL"
        ))
    }

    private fun addResponseInfoWhenSimulert(simulertPensjonList: List<SimulertPensjon>, stillingsprosentResponse: StillingsprosentResponse): List<SimulertPensjon> {
        val utelatteTpNr = stillingsprosentResponse.exceptions
                .map(ExecutionException::cause)
                .filterIsInstance<StillingsprosentCallableException>()
                .map(StillingsprosentCallableException::tpOrdning)
                .map(TPOrdning::tpId)
        val inkluderteTpNr = stillingsprosentResponse.tpOrdningStillingsprosentMap.keys
                .map(TPOrdning::tpId)
        return simulertPensjonList.onEach { simulertPensjon ->
            if(simulertPensjon is SimulertPensjonOK) {
                simulertPensjon.utelatteTpnr = utelatteTpNr
                simulertPensjon.inkluderteTpnr = inkluderteTpNr
                if (utelatteTpNr.isNotEmpty())
                    simulertPensjon.status = "UFUL"
            }
        }.also {
            incrementMetrics(it, utelatteTpNr)
        }
    }

    private fun incrementMetrics(simulertPensjonList: List<SimulertPensjon>, utelatteTpNr: List<String?>) {
        val ufullstendig = utelatteTpNr.isNotEmpty()
        val mangelfull = simulertPensjonList.filterIsInstance<SimulertPensjonOK>().flatMap(SimulertPensjonOK::utbetalingsperioder).any { it == null }
        if (ufullstendig)
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_UFUL)
        if (mangelfull)
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_MANGEL)
        if (!ufullstendig && !mangelfull)
            metrics.incrementCounter(APP_NAME, APP_TOTAL_SIMULERING_OK)
    }

    private fun getTpLeverandorer(tpOrdningList: List<TPOrdning>) =
            asyncExecutor.executeAsync(
                    tpOrdningList.map { tpOrdning ->
                        tpOrdning to FindTpLeverandorCallable(tpOrdning, tpConfigConsumer, tpLeverandorList)
                    }.toMap()
            ).resultMap

    companion object {
        @JvmStatic
        @Suppress("JAVA_CLASS_ON_COMPANION")
        private val LOG = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

}